network/src/connectivity_manager/mod.rs

```diff
@@ -394,108 +406,171 @@ where
         }
     }
 
-    async fn dial_eligible_peers<'a>(
+    fn dial_eligible_peers<'a>(
         &'a mut self,
         pending_dials: &'a mut FuturesUnordered<BoxFuture<'static, PeerId>>,
     ) {
-        let eligible = self.eligible.read().clone();
+        let to_connect = self.choose_peers_to_dial();
+        for (peer_id, peer) in to_connect {
+            self.queue_dial_peer(peer_id, peer, pending_dials);
+        }
+    }
+
+    fn choose_peers_to_dial(&mut self) -> Vec<(PeerId, DiscoveredPeer)> {
         let to_connect: Vec<_> = self
             .discovered_peers
             .0
             .iter()
             .filter(|(peer_id, peer)| {
-                eligible.contains_key(peer_id)  // The node is eligible to be dialed.
+                peer.is_eligible_to_be_dialed() // The node is eligible to dial
                     && self.connected.get(peer_id).is_none() // The node is not already connected.
                     && self.dial_queue.get(peer_id).is_none() // There is no pending dial to this node.
-                    && !peer.addrs.is_empty() // There is an address to dial.
             })
-            .map(|(peer_id, peer)| (peer_id, &peer.addrs))
             .collect();
 
+        let max_peers_to_connect = to_connect.len();
+
+        // Sort by role for choosing
+        let mut by_role: HashMap<_, Vec<_>> = HashMap::new();
+        for (peer_id, peer) in to_connect {
+            by_role.entry(peer.role).or_default().push((peer_id, peer));
+        }
+
         // Limit the number of dialed connections from a Full Node
         // This does not limit the number of incoming connections
         // It enforces that a full node cannot have more outgoing connections than `connection_limit`
         // including in flight dials.
         let to_connect_size = if let Some(conn_limit) = self.outbound_connection_limit {
             min(
                 conn_limit.saturating_sub(self.connected.len() + self.dial_queue.len()),
-                to_connect.len(),
+                max_peers_to_connect,
             )
         } else {
-            to_connect.len()
+            max_peers_to_connect
+        };
+
+        let network_id = self.network_context.network_id();
+        let roles_to_choose = match network_id {
+            NetworkId::Validator => vec![PeerRole::Validator],
+            NetworkId::Private(_) => {
+                if network_id.is_vfn_network() {
+                    vec![PeerRole::ValidatorFullNode]
+                } else {
+                    // TODO: This is not a well defined network
+                    vec![PeerRole::PreferredUpstream, PeerRole::Upstream]
+                }
+            }
+            NetworkId::Public => vec![
+                PeerRole::PreferredUpstream,
+                PeerRole::ValidatorFullNode,
+                PeerRole::Upstream,
+            ],
         };
 
+        let mut chosen_peers: Vec<(&PeerId, &DiscoveredPeer)> = Vec::new();
```

Comment:
can avoid the extra stuff at the bottom if this is just `Vec<(PeerId, DiscoveredPeer)>`